### DISCUSSION TITLE
HttpURLConnection instrumentation doesn't close spans for a lot of scenarios

### DISCUSSION THREAD

**Discussion by surbhiia** [open-telemetry/opentelemetry-java-instrumentation#9869] [Q&A]
Had a question on HttpURLConnection instrumentation - The instrumentation [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/c973b9ebd41bfb1f3dd046f054dddcd8d0d32a94/instrumentation/http-url-connection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/httpurlconnection/HttpUrlConnectionInstrumentation.java#L125) does not close spans unless an exception occurs or getInputStream is called. It seems to miss a lot of other scenarios like for example:

//Example 1: No exception occurs and getInputStream is not called
URL url = new URL("http://example.com/”);
URLConnection urlConnection = url.openConnection();
Map<String, List<String>> map = urlConnection.getHeaderFields();
// Or any other similar api’s other than getInputStream that read from connection are called

//Example 2: Post scenario
URL url = new URL("http://example/.com”);
URLConnection urlConnection = url.openConnection();
urlConnection.setDoOutput(true);
OutputStreamWriter out = new OutputStreamWriter(urlConnection.getOutputStream());
out.write("Output test content");
out.close()

Am I wrong in my understanding Or was it discussed before that this was not meant to cover all such possible scenarios?

I found a similar issue that touched this before [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2097) but my concern is more from having a complete instrumentation (that covers all possible general scenarios) perspective.

**Comment by trask** [comment]
hi @surbhiia! IIRC `getInputStream()` gets called internally by other `HttpURLConnection` methods. Have you tested the above examples using the Java agent?

**Reply by surbhiia** [reply]
I haven't tested this yet. But from my understanding getInputStream is only called internally from getResponseCode API. And also from getResponseMessage which calls getResponseCode. 

But it could be the case that not visible to us (intrinsically) all apis are calling getInputStream somewhere! I will test if above scenarios create spans and get back to you.

**Comment by breedx-splk** [comment]
Related: https://github.com/open-telemetry/opentelemetry-android/pull/133#issuecomment-1813328503
